### PR TITLE
stm32 OSPI flash driver enables XiP on external NOR flash

### PR DESCRIPTION
**Pull request by FRASTM**
With this PR the stm32u5 and stm32h7 target board with octo-NOR flash can configure this external flash to eXecute In Place
a sample code application stored on this external memory.
The ospi flash driver is configuring the external NOR in MemoryMapped mode at init so that code is XiP at the external NOR flash address.
This is demonstrated by the samples/application_development/code_relocation_nocopy/ running on the stm32u585 or stm32h7b3 disco kits.

For stm32h7 serie, this PR requires the https://github.com/zephyrproject-rtos/zephyr/pull/68593  
  and include a sub-region  of the EXTMEM where code is executed in XiP where
     `zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_IO) )>;`

For stm32u5 serie this PR requires the https://github.com/zephyrproject-rtos/zephyr/pull/68590

Requires the PR https://github.com/zephyrproject-rtos/zephyr/pull/68377 to flash correclty the code to the external flash with west command.

This PR is replacing the https://github.com/zephyrproject-rtos/zephyr/pull/61082

### CODE REVIEW

**Comment by FRASTM** [comment]
Running the sample to relocate code on the external NOR flash of the b_u585i_iot02a disco kit:
` west build -p always -b b_u585i_iot02a samples/application_development/code_relocation_nocopy/
`
gives 
```
*** Booting Zephyr OS build zephyr-v3.5.0-5280-g36d125514e1f ***
Address of main function 0x8000599
Address of function_in_ext_flash 0x70000001
Address of var_ext_sram_data 0x20000080 (10)
Address of function_in_sram 0x20000001                           
Address of var_sram_data 0x20000084 (10)                         
Hello World! b_u585i_iot02a         
```

**Comment by GeorgeCGV** [comment]
@FRASTM we started to experience application freezing when it is started by the bootloader (BL).

 The H7 freezes in the `SCB_DisableDCache` when `CONFIG_NO_OPTIMIZATIONS=y`. That happens within `z_arm_reset` in `z_arm_init_arch_hw_at_boot` when it tries to reset the D-CACHE due to it being enabled in BL.

A solution is to set `CONFIG_CACHE_MANAGEMENT=n` in the app, but leave it on in the bootloader or vice versa.

Worked fine before the introduction of `CONFIG_CACHE_MANAGEMENT`.

Did you experience something like that on H7 while working on the XIP support?

**Comment by FRASTM** [comment]
> @FRASTM started to experience application freezing when it was started by the bootloader (BL).
> 
> The H7 application freezes in the `SCB_DisableDCache` when `CONFIG_NO_OPTIMIZATIONS=y`. That happens within `z_arm_reset` in `z_arm_init_arch_hw_at_boot` when it tries to reset the D-CACHE due to it being enabled in BL.
> 
> A solution is to set `CONFIG_CACHE_MANAGEMENT=n` in the app, but leave it on in the bootloader or vice versa.
> 
> Everything worked fine before the introduction of `CONFIG_CACHE_MANAGEMENT`.
> 
> Did you experience something like that on H7 while working on the XIP support?
> 
> Note: what is interesting is that it works with `CONFIG_NO_OPTIMIZATIONS=n`

Yes, I had some pb with the stm32h7 cache, on my side, for that, I changed the memory-attribute of the ext RAM section
```
diff --git a/dts/arm/st/h7/stm32h7.dtsi b/dts/arm/st/h7/stm32h7.dtsi
index 64ccfbf724..2351eb0e49 100644
--- a/dts/arm/st/h7/stm32h7.dtsi
+++ b/dts/arm/st/h7/stm32h7.dtsi
@@ -45,11 +45,12 @@
 		};
 	};
 
-	quadspi_memory: memory@90000000 {
-		compatible = "zephyr,memory-region", "mmio-sram";
-		reg = <0x90000000 DT_SIZE_M(256)>;
-		zephyr,memory-region = "QSPI";
-		zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_EXTMEM) )>;
+	ext_memory: memory@90000000 {
+		compatible = "zephyr,memory-region";
+		reg = <0x90000000 DT_SIZE_M(64)>; /* 512 Mbits */
+		zephyr,memory-region = "EXTMEM";
+		/* The ATTR_MPU_EXTMEM attribut causing a MPU FAULT */
+		zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_IO) )>;
 	};
 
 	clocks {
```

**Comment by GeorgeCGV** [comment]
> > @FRASTM started to experience application freezing when it was started by the bootloader (BL).
> > The H7 application freezes in the `SCB_DisableDCache` when `CONFIG_NO_OPTIMIZATIONS=y`. That happens within `z_arm_reset` in `z_arm_init_arch_hw_at_boot` when it tries to reset the D-CACHE due to it being enabled in BL.
> > A solution is to set `CONFIG_CACHE_MANAGEMENT=n` in the app, but leave it on in the bootloader or vice versa.
> > Everything worked fine before the introduction of `CONFIG_CACHE_MANAGEMENT`.
> > Did you experience something like that on H7 while working on the XIP support?
> > Note: what is interesting is that it works with `CONFIG_NO_OPTIMIZATIONS=n`
> 
> Yes, I had some pb with the stm32h7 cache, on my side, for that, I changed the memory-attribute of the ext memory section
> 
> ```diff
> diff --git a/dts/arm/st/h7/stm32h7.dtsi b/dts/arm/st/h7/stm32h7.dtsi
> index 64ccfbf724..2351eb0e49 100644
> --- a/dts/arm/st/h7/stm32h7.dtsi
> +++ b/dts/arm/st/h7/stm32h7.dtsi
> @@ -45,11 +45,12 @@
>  		};
>  	};
>  
> -	quadspi_memory: memory@90000000 {
> -		compatible = "zephyr,memory-region", "mmio-sram";
> -		reg = <0x90000000 DT_SIZE_M(256)>;
> -		zephyr,memory-region = "QSPI";
> -		zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_EXTMEM) )>;
> +	ext_memory: memory@90000000 {
> +		compatible = "zephyr,memory-region";
> +		reg = <0x90000000 DT_SIZE_M(64)>; /* 512 Mbits */
> +		zephyr,memory-region = "EXTMEM";
> +		/* The ATTR_MPU_EXTMEM attribut causing a MPU FAULT */
> +		zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_IO) )>;
>  	};
>  
>  	clocks {
> ```

The issue in my case is not related to MPU. It can easily be observed with any region attribute.

When
* bootloader is built it enables the D-Cache and I-Cache as long as `CONFIG_CACHE_MANAGEMENT=y`.
* application  is built with `CONFIG_CACHE_MANAGEMENT=y` in addition to
```
CONFIG_DEBUG=y
CONFIG_NO_OPTIMIZATIONS=y
```

Eventually, when Bl jumps to the application and Zephyr calls `z_arm_init_arch_hw_at_boot` it detects that the D-Cache is enabled and tries to disable it. The disabling `SCB_DisableDCache` loops forever. I am not sure why it is only observable when `CONFIG_NO_OPTIMIZATIONS=y`.

**Comment by benediktibk** [comment]
> @benediktibk shouldn't
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/aef6588820d07d79a00ea5d0a5818d89f2c1d066/arch/arm/core/cortex_m/scb.c#L116-L146

I don't think so, because the whole idea behind z_arm_init_arch_hw_at_boot is to get the system after after a bootloader into the same state as it is after a full reset. Starting from this then the caches are activated if they are enabled in the config of the actual application.

The big question for me is why SCB_DisableDCache seems to get stuck, because it should be safe to call it at this point? From the documentation it is not completely clear if it is required to call SCB_CleanDCache beforehand, but it might be worth a try?

**Comment by GeorgeCGV** [comment]
@benediktibk the issues seems to come from https://github.com/ARM-software/CMSIS_5/issues/620. Currently, CMSIS v5 is used. The upstream fixes the issue in develop branch v6 which is not released yet and CMSIS v6 (_not used_). It could be great to have a mention of the issue in the docs.

**Comment by FRASTM** [comment]
rebase on db1a4a61be6d7fc277f0ae519c0228af7d9d297b
minimal ospi configuration to run the samples/application_development/code_relocation_nocopy/  on b_u585i_iot02a and stm32h7b3i_dk disco boards

For flashing (west flash) the stm32h7b3i_dk disco kit, the https://github.com/zephyrproject-rtos/zephyr/pull/68377 is required
```
*** Booting Zephyr OS build v3.6.0-3798-gce5ed69f86bd ***
Address of main function 0x8000685
Address of function_in_ext_flash 0x90000001
Address of var_ext_sram_data 0x240000e0 (10)
Address of function_in_sram 0x24000001
Address of var_sram_data 0x240000e4 (10)
Hello World! stm32h7b3i_dk
```

**Review by erwango — DISMISSED** [review]

File: samples/application_development/code_relocation_nocopy/README.rst [hunk 1]
```diff
@@ -51,3 +51,29 @@ Execution output
...
+***********************************************
+
+The b_u585i_iot02a has 64MB of external flash attached via OSPI. It is mapped
+to 0x70000000.
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/application_development/code_relocation_nocopy
+   :board: b_u585i_iot02a
+   :goals: build flash
+   :compact:
```
Comment (erwango) [inline samples/application_development/code_relocation_nocopy/README.rst:66]: Move this section before `Execution output`

File: samples/application_development/code_relocation_nocopy/README.rst [hunk 2]
```diff
@@ -51,3 +51,29 @@ Execution output
...
+
+.. code-block:: console
+
+  *** Booting Zephyr OS build v3.6.0-3797-g2aae7ba69ae6 ***
+  Address of main function 0x8000599
+  Address of function_in_ext_flash 0x70000001
+  Address of var_ext_sram_data 0x20000080 (10)
+  Address of function_in_sram 0x20000001
+  Address of var_sram_data 0x20000084 (10)
+  Hello World! b_u585i_iot02a
```
Comment (erwango) [inline samples/application_development/code_relocation_nocopy/README.rst:79]: Don't need to be duplicated

**Review by erwango — CHANGES_REQUESTED** [review]

File: samples/application_development/code_relocation_nocopy/README.rst (hunk 2 above)
Comment (FRASTM) [reply samples/application_development/code_relocation_nocopy/README.rst]: section canceled

File: samples/application_development/code_relocation_nocopy/README.rst (hunk 1 above)
Comment (FRASTM) [reply samples/application_development/code_relocation_nocopy/README.rst:52]: section moved

**Review by erwango — DISMISSED** [review]

**Review by gautierg-st — DISMISSED** [review]

File: drivers/flash/flash_stm32_ospi.c [hunk 3]
```diff
@@ -978,11 +978,149 @@ static int stm32_ospi_mem_reset(const struct device *dev)
...
+static bool stm32_ospi_is_memorymap(const struct device *dev)
+{
+	struct flash_stm32_ospi_data *dev_data = dev->data;
+
+	return ((READ_BIT(dev_data->hospi.Instance->CR,
+			  OCTOSPI_CR_FMODE) == OCTOSPI_CR_FMODE) ?
+			  true : false);
+}
+
+/* Abort the Memorymapped Mode so that erasing is possible */
```
Comment (arbrauns) [inline drivers/flash/flash_stm32_ospi.c:1103]: This is an inaccurate description of `HAL_OSPI_Abort` - it's not the memory-mapped mode that's aborted, it's the idle SPI transaction from the last memory-mapped read (unless `OCTOSPI_LPTR` is set, chip-select is never automatically deasserted by hardware if memory-mapped mode is disabled). Memory-mapped mode is still active after an abort.

This also makes many comments added in this file incorrect/inaccurate.

**Review by arbrauns — CHANGES_REQUESTED** [review]

File: drivers/flash/flash_stm32_ospi.c (hunk 3 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c:1103]: Ok,
 changing the stm32_ospi_abort function  the same way as flash_stm32_qspi driver

**Comment by FRASTM** [comment]
fixing build errors
Can pass the   samples/application_development/code_relocation_nocopy/  on the **b_u585i_iot02a** target when the external octo-NOR is in MemoryMapped mode for XiP: 

```
*** Booting Zephyr OS build v3.6.0-4481-gc78c2550c34e ***                       
Address of main function 0x8000599                                              
Address of function_in_ext_flash 0x70000001                                     
Address of var_ext_sram_data 0x20000080 (10)                                    
Address of function_in_sram 0x20000001                                          
Address of var_sram_data 0x20000084 (10)                                        
Hello World! b_u585i_iot02a    
```

**Comment by ajarmouni-st** [comment]
@FRASTM You should rebase on top of d490616d08c1799f7e314a2663da19180f6b6123 instead of adding the commits of https://github.com/zephyrproject-rtos/zephyr/pull/68377 on top of this PR's branch.

**Comment by FRASTM** [comment]
> @FRASTM You should rebase on top of [d490616](https://github.com/zephyrproject-rtos/zephyr/commit/d490616d08c1799f7e314a2663da19180f6b6123) instead of adding the commits of #68377 on top of this PR's branch.

done

**Comment by FRASTM** [comment]
- Rebase on c1400536548bb861c630f81d8aced864f1a85ca1
- Add a commit to remove the stm32h5 serie (which now is using the flash_stm32_xspi_ driver) from the flash_stm32_ospi driver code
- samples/application_development/code_relocation_nocopy/   can executes code (XiP) from the external flash. STM32CubeProgrammer with external loader is used to flash to the disco board

**Review by ajarmouni-st — DISMISSED** [review]

**Review by erwango — COMMENTED** [review]
Otherwise LGTM

File: drivers/flash/flash_stm32_ospi.c [hunk 4]
```diff
@@ -1286,6 +1304,13 @@ static int flash_stm32_ospi_read(const struct device *dev, off_t addr,
...
+	/* If not MemMapped then configure it */
+	if (!stm32_ospi_is_memorymap(dev)) {
+		if (stm32_ospi_set_memorymap(dev) != 0) {
+			LOG_ERR("READ failed: cannot enable MemoryMap");
+			return -EIO;
+		}
+	}
 	if (stm32_ospi_is_memorymap(dev)) {
 		LOG_DBG("MemoryMapped Read offset: 0x%lx, len: %zu",
 			(long)(STM32_OSPI_BASE_ADDRESS + addr),
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: The rational was to keep  the sequence as similar as possible with  the **flash_stm32_qspi_read**

[the remark this answers is not in the document]
Comment (erwango) [reply drivers/flash/flash_stm32_ospi.c]: ok, but in `flash_stm32_qspi_read` there is not second `if` neither actually (which is even better than doing the `else` I proposed)

[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: removing the sequence the test on stm32_ospi_is_memorymap(dev) : if memory mapped mode is not, we set : at line1314, the memory mapped mode is always set.

**Review by erwango — DISMISSED** [review]

**Review by ajarmouni-st — DISMISSED** [review]

**Comment by FRASTM** [comment]
fixing CI error

**Review by ajarmouni-st — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
@de-nordic Ready for you review

**Review by de-nordic — APPROVED** [review]

File: drivers/flash/flash_stm32_ospi.c [hunk 5]
```diff
@@ -978,11 +978,149 @@ static int stm32_ospi_mem_reset(const struct device *dev)
...
+
+	/* Initialize the program command */
+	s_command.OperationType = HAL_OSPI_OPTYPE_WRITE_CFG;
+	if (dev_cfg->data_rate == OSPI_STR_TRANSFER) {
+		s_command.Instruction = (dev_cfg->data_mode == OSPI_SPI_MODE)
+					? ((stm32_ospi_hal_address_size(dev) ==
+					HAL_OSPI_ADDRESS_24_BITS)
+						? SPI_NOR_CMD_PP
+						: SPI_NOR_CMD_PP_4B)
+					: SPI_NOR_OCMD_PAGE_PRG;
```
Comment (GeorgeCGV) [inline drivers/flash/flash_stm32_ospi.c:1068]: Doesn't make use of set `write_opcode`. Won't work for some configurations.

File: drivers/flash/flash_stm32_ospi.c [hunk 6]
```diff
@@ -978,11 +978,149 @@ static int stm32_ospi_mem_reset(const struct device *dev)
...
+	/* Configure octoflash in MemoryMapped mode */
+	if ((dev_cfg->data_mode == OSPI_SPI_MODE) &&
+		(stm32_ospi_hal_address_size(dev) == HAL_OSPI_ADDRESS_24_BITS)) {
+		/* OPI mode and 3-bytes address size not supported by memory */
+		LOG_ERR("OSPI_SPI_MODE in 3Bytes addressing is not supported");
+		return -EIO;
+	}
+
+	/* Initialize the read command */
+	s_command.OperationType = HAL_OSPI_OPTYPE_READ_CFG;
```
Comment (GeorgeCGV) [inline drivers/flash/flash_stm32_ospi.c:1000]: Will require some modification to make use of `read_opcode`, `read_dummy_cycles` in non DTR, OPI modes.

<details>

<summary>Example of a separate read cmd preparation function</summary>

```c


static int ospi_prepare_read_s_command(const struct device *dev, OSPI_Regulars_commandTypeDef *s_command)
{
	struct flash_stm32_ospi_data *dev_data = dev->data;
	const struct flash_stm32_ospi_config *dev_cfg = dev->config;

	s_command->OperationType = HAL_OSPI_OPTYPE_READ_CFG;
	s_command->FlashId = HAL_OSPI_FLASH_ID_1;
	/* default, may be overriden later */
	s_command->AlternateBytesMode = HAL_OSPI_ALTERNATE_BYTES_NONE;
	s_command->SIOOMode = HAL_OSPI_SIOO_INST_EVERY_CMD;

	if (dev_cfg->data_rate == OSPI_DTR_TRANSFER) {
		s_command->InstructionDtrMode = HAL_OSPI_INSTRUCTION_DTR_ENABLE;
		s_command->AddressDtrMode = HAL_OSPI_ADDRESS_DTR_ENABLE;
		s_command->DataDtrMode = HAL_OSPI_DATA_DTR_ENABLE;
		s_command->DQSMode = HAL_OSPI_DQS_ENABLE;
	}

	if (dev_cfg->data_mode == OSPI_OPI_MODE) {
		s_command->AddressSize = HAL_OSPI_ADDRESS_32_BITS;
		s_command->InstructionSize = HAL_OSPI_INSTRUCTION_16_BITS;

		s_command->InstructionMode = HAL_OSPI_INSTRUCTION_8_LINES;
		s_command->AddressMode = HAL_OSPI_ADDRESS_8_LINES;
		s_command->DataMode = HAL_OSPI_DATA_8_LINES;
	} else {
		s_command->AddressSize = stm32_ospi_hal_address_size(dev);
		s_command->InstructionSize = HAL_OSPI_INSTRUCTION_8_BITS;

		switch (dev_data->read_mode) {
		case JESD216_MODE_112: {
			s_command->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
			s_command->AddressMode = HAL_OSPI_ADDRESS_1_LINE;
			s_command->DataMode = HAL_OSPI_DATA_2_LINES;
			break;
		}
		case JESD216_MODE_122: {
			s_command->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
			s_command->AddressMode = HAL_OSPI_ADDRESS_2_LINES;
			s_command->DataMode = HAL_OSPI_DATA_2_LINES;
			break;
		}
		case JESD216_MODE_114: {
			s_command->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
			s_command->AddressMode = HAL_OSPI_ADDRESS_1_LINE;
			s_command->DataMode = HAL_OSPI_DATA_4_LINES;
			break;
		}
		case JESD216_MODE_144: {
			s_command->InstructionMode = HAL_OSPI_INSTRUCTION_1_LINE;
			s_command->AddressMode = HAL_OSPI_ADDRESS_4_LINES;
			s_command->DataMode = HAL_OSPI_DATA_4_LINES;
			break;
		}
		default:
			if (dev_cfg->data_mode == OSPI_QUAD_MODE) {
				cmd->InstructionMode = HAL_OSPI_INSTRUCTION_4_LINES;
				cmd->AddressMode = HAL_OSPI_ADDRESS_4_LINES;
				cmd->DataMode = HAL_OSPI_DATA_4_LINES;
				break;
			} else if (dev_cfg->data_mode == OSPI_DUAL_MODE) {
				cmd->InstructionMode = HAL_OSPI_INSTRUCTION_2_LINES;
				cmd->AddressMode = HAL_OSPI_ADDRESS_2_LINES;
				cmd->DataMode = HAL_OSPI_DATA_2_LINES;
				break;
			} else {
				return -ENOTSUP;
			}
		}
	}

	if (dev_cfg->data_rate == OSPI_DTR_TRANSFER) {
		/* DTR transfer rate (==> Octal mode) */
		s_command->Instruction = SPI_NOR_OCMD_DTR_RD;
		s_command->DummyCycles = SPI_NOR_DUMMY_RD_OCTAL_DTR;
	} else {
		/* STR transfer rate */
		if (dev_cfg->data_mode == OSPI_OPI_MODE) {
			/* OPI and STR */
			s_command->Instruction = SPI_NOR_OCMD_RD;
			s_command->DummyCycles = SPI_NOR_DUMMY_RD_OCTAL;
		} else {
			/* use SFDP:BFP read instruction */
			s_command->Instruction = dev_data->read_opcode;
			s_command->DummyCycles = dev_data->read_dummy_cycles;
		}
	}

	return 0;
}
```

</details>

File: drivers/flash/flash_stm32_ospi.c (hunk 5 above)
Comment (erwango) [reply drivers/flash/flash_stm32_ospi.c:1068]: @GeorgeCGV, thanks! Let's make it a follow up PR

File: drivers/flash/flash_stm32_ospi.c (hunk 6 above)
Comment (erwango) [reply drivers/flash/flash_stm32_ospi.c:1000]: Ditto

Comment (GeorgeCGV) [reply drivers/flash/flash_stm32_ospi.c:1000]: Opened #73840.

**Pull request merged**